File: app/src/main/java/com/bubelov/coins/dagger/DatabaseModule.java
package com.bubelov.coins.dagger;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.preference.PreferenceManager;

import com.bubelov.coins.database.DbHelper;
import com.bubelov.coins.service.DatabaseSync;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * @author Igor Bubelov
 */

@Module
public class DatabaseModule {
    @Provides
    @Singleton
    SQLiteDatabase database(Context context) {
        SQLiteOpenHelper helper = new DbHelper(context);
        return helper.getWritableDatabase();
    }

    @Provides
    @Singleton
    DatabaseSync databaseSync() {
        return new DatabaseSync();
    }

    @Provides
    @Singleton
    SharedPreferences preferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }
}

File: app/src/main/java/com/bubelov/coins/dagger/DatabaseModule.kt
package com.bubelov.coins.dagger

import android.arch.persistence.room.Room
import android.content.Context
import com.bubelov.coins.db.Database
import com.bubelov.coins.db.DatabaseConfig
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * @author Igor Bubelov
 */

@Module
class DatabaseModule {
    @Provides
    @Singleton
    fun provideDatabase(context: Context, databaseConfig: DatabaseConfig) = Room.databaseBuilder(context, Database::class.java, "db.sqlite3").apply {
        if (databaseConfig.canUseMainThread) {
            allowMainThreadQueries()
        }
    }.build()

    @Provides
    @Singleton
    fun provideDatabaseConfig() = DatabaseConfig(canUseMainThread = false)
}